### PR TITLE
feat(rules): mutual-exclusion audience rule (bucketed_into_experience_key)

### PR DESCRIPTION
**Pull request by abbaseya**
## Summary

Adds a declarative mutual-exclusion audience rule to the PHP SDK: a new rule type **`bucketed_into_experience_key`** whose value is a **target experience key**. It matches iff the visitor is already bucketed into that experience per the SDK's stored bucketing state (in-memory merged with the cache/DataStore), and combines with `negated: true` to express "NOT in experiment X". **Zero new application inputs** — it works with empty visitor properties.

Inert until served configs carry the new rule type (backend registration is a separate spec). One of six sibling SDK specs; consistency is defined by the shared normative contract + the identical 8-row fixture (parity oracle: JS SDK PR convertcom/javascript-sdk#416).

> **Stacked PR** — base is `feat/experiment-preview` (qs-02), **not** `main`. Do not merge before the base.

## Design

- **RuleManager is byte-unchanged** (strongest AC7 guarantee). Resolution happens DataManager-side via a **synthetic-delegation seam** in `filterMatchedRecordsWithRule`: resolve `bucketedRaw` read-only, then route a synthetic single-key `equals` rule through the unmodified `isRuleMatched()` so negation reuses the real `Comparisons::returnNegationCheck`.
- The PHP-specific empty-visitor-properties gate in `matchRulesByField` (`if ($visitorProperties)`) is widened to `if ($visitorProperties || $hasBucketingExclusionAudience)`, triggered **only** when an audience's rule tree is a sole `bucketed_into_experience_key` element — generic-only audiences keep byte-identical behavior.
- Resolution: `target = getEntity(value,'experiences')`; `bucketedRaw = target !== null && array_key_exists((string)target.id, getData(visitorId).bucketing)`; `matched = negated ? !bucketedRaw : bucketedRaw`. Read-only (no target bucketing, no store writes, no tracking). Unknown target → `false` + warn naming the key.

## Acceptance criteria — evidence

| AC | What | Evidence |
|----|------|----------|
| AC1 | 8-row fixture, single data provider | `MutualExclusionRuleResolutionTest` — `@DataProviderExternal` over `MutualExclusionFixture::rows()`, 8/8 |
| AC2 | End-to-end exclusion | `MutualExclusionIntegrationTest` — run A → buckets; run B (negated audience, ALL) → excluded; fresh visitor buckets B |
| AC3 | Persistence (new instance) | `MutualExclusionPersistenceTest` — decision written via DataStore in instance #1 excludes B in a NEW instance (fixture row 8) |
| AC4 | No new inputs (empty props) | all resolution + integration cases run with `visitorProperties=[]`; the :410 gate is widened for the exclusion rule |
| AC5 | Read-only | `MutualExclusionReadOnlyTest` — spies assert zero target bucketing, zero store writes, zero tracking events |
| AC6 | ALL / ANY combination | `MutualExclusionCombinationTest` — exclusion + generic rule under ALL and ANY |
| AC7 | Generic-rule regression lock | `MutualExclusionGenericRegressionTest` (9 tests) + full suite green; generic text/numeric/bool bit-identical (visitorId passed and null); generic-only + empty props still excluded |
| AC8 | Unknown-target warning | rows 6/7 assert a warn log naming the unresolved key (`Messages::BUCKETING_EXCLUSION_TARGET_NOT_FOUND`) |

## Tests

Full php-sdk suite **1168 tests / 4439 assertions green** (34 pre-existing env skips). New mutual-exclusion suites: **33 tests / 138 assertions**. `phpstan` (level 6, configured `*/src` scope) clean; `php-cs-fixer` clean. `RuleManager.php`, `DataManagerInterface.php`, `Comparisons.php`, `Types/**` byte-unchanged.

## Files

Production: `packages/Data/src/DataManager.php` (seam + 4 private helpers), `packages/Enums/src/RuleType.php` (new backed enum), `packages/Enums/src/Messages.php` (warn constant). Remainder: tests + fixture + config.

## Notes / carry-forward

- Cross-request exclusion requires a persistent cache/DataStore backing (Redis/Memcached); the default in-memory ArrayCache is request-scoped.
- Old SDK versions receiving this rule fail closed (evaluate `false`, negation unapplied) — the guide should recommend `ALL` and state the minimum SDK version.
- A single audience carrying **both** a generic and an exclusion rule in one tree is out of scope (no producer exists yet); the exclusion element fails closed in that case. The backend-registration sibling spec must honor "exclusion rule is the sole rule in its audience".
- "Zero visitor inputs" still requires location matching to be satisfied (tests set `ignoreLocationProperties: true`); worth a release-guide note.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/Data/src/DataManager.php [hunk 1]
```diff
@@ -418,7 +432,8 @@ public function matchRulesByField(
                         $audiencesToCheck,
                         $visitorProperties,
```
[the remark this answers is not in the document]
Comment (abbaseya) [reply packages/Data/src/DataManager.php]: Valid — fixed in ad15917. Under `strict_types=1` the qs-03-widened gate (`if ($visitorProperties || $hasBucketingExclusionAudience)`) can reach this call with `$visitorProperties` still `null` when the caller supplies no properties at all (`BucketingAttributes::setVisitorProperties(?array)`), throwing a TypeError against the non-nullable `array` param. Coerced to `$visitorProperties ?? []` — safe because the exclusion path reads stored bucketing state via `getData()`, not property content. Also added a regression test (`testNullVisitorPropertiesDoesNotThrowOnNegatedExclusion`) covering the genuine null-caller path for AC4, which the prior `[]`-only tests had masked.

File: packages/Data/src/DataManager.php [hunk 2]
```diff
@@ -1402,6 +1448,120 @@ public function filterMatchedRecordsWithRule(
...
+     * DataManager.php empty-visitorProperties audience-evaluation gate SOLELY
+     * for experiences whose audience tree needs this rule type (AC4), while
+     * leaving the gate untouched for every generic-only audience (AC7).
+     *
+     * @param array<int, array<string, mixed>> $audiences
+     */
+    private function _audiencesContainBucketedIntoExperienceKeyRule(array $audiences): bool
+    {
+        foreach ($audiences as $audience) {
+            if (!empty($audience['rules']) && $this->_findSoleBucketedIntoExperienceKeyRule($audience['rules']) !== null) {
```
[the remark this answers is not in the document]
Comment (abbaseya) [reply packages/Data/src/DataManager.php]: Valid — fixed in ad15917. Added an `is_array()` guard here and at the sibling qs-03 call site in `filterMatchedRecordsWithRule()` (both feed the strict `array`-typed `_findSoleBucketedIntoExperienceKeyRule()`), matching the existing `is_array($experience['audiences'])` guard at line 408.

File: packages/Data/src/DataManager.php [hunk 3]
```diff
@@ -1402,6 +1448,120 @@ public function filterMatchedRecordsWithRule(
...
+        if ($target === null) {
+            $this->_loggerManager?->warn(
+                'DataManager.filterMatchedRecordsWithRule()',
+                str_replace('#', $targetKey, Messages::BUCKETING_EXCLUSION_TARGET_NOT_FOUND)
+            );
+            $bucketedRaw = false;
+        } else {
+            $visitorData = $this->getData($visitorId) ?? [];
+            $bucketingData = $visitorData['bucketing'] ?? [];
+            $bucketedRaw = array_key_exists((string)$target['id'], $bucketingData);
```
[the remark this answers is not in the document]
Comment (abbaseya) [reply packages/Data/src/DataManager.php]: Valid — fixed in ad15917. Changed to `(string)($target['id'] ?? '')`, consistent with the same method's `(string)($rule['value'] ?? '')`. This matters here because PHPUnit runs with `failOnWarning=true`, so the E_WARNING would fail the suite.

**Review by DmytroConvert — APPROVED** [review]
